components/CalEPD/models/gdeh0154z90.cpp

```diff
@@ -0,0 +1,246 @@
+#include "gdeh0154z90.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include "esp_log.h"
+#include "freertos/task.h"
+
+// Partial Update Delay
+#define GDEH0154Z90_PU_DELAY 300
+
+// Constructor
+Gdeh0154z90::Gdeh0154z90(EpdSpi &dio) : Adafruit_GFX(GDEH0154Z90_WIDTH, GDEH0154Z90_HEIGHT),
+                                        Epd(GDEH0154Z90_WIDTH, GDEH0154Z90_HEIGHT), IO(dio)
+{
+    ESP_LOGI(TAG, "Gdeh0154z90() %d*%d\n", GDEH0154Z90_WIDTH, GDEH0154Z90_HEIGHT);
+}
+
+//Initialize the display
+void Gdeh0154z90::init(bool debug)
+{
+    debug_enabled = debug;
+    if (debug_enabled)
+    {
+        ESP_LOGI(TAG, "Gdeh0154z90::init(%d)", debug);
+    }
+    IO.init(4, debug); // 4MHz frequency
+
+    ESP_LOGI(TAG, "Free heap:%d", xPortGetFreeHeapSize());
+    fillScreen(EPD_WHITE);
+}
+
+void Gdeh0154z90::fillScreen(uint16_t color)
+{
+    uint8_t black = GDEH0154Z90_8PIX_WHITE;
+    uint8_t red = GDEH0154Z90_8PIX_RED_WHITE;
+
+    if (color == EPD_BLACK)
+    {
+        black = GDEH0154Z90_8PIX_BLACK;
+    }
+    else if (color == EPD_RED)
+    {
+        red = GDEH0154Z90_8PIX_RED;
+    }
+
+    for (uint16_t x = 0; x < sizeof(_black_buffer); x++)
+    {
+        _black_buffer[x] = black;
+        _red_buffer[x] = red;
+    }
+
+    if (debug_enabled)
+    {
+        ESP_LOGI(TAG, "fillScreen(%d) _buffer len:%d", color, sizeof(_black_buffer));
+    }
+}
+
+void Gdeh0154z90::drawPixel(int16_t x, int16_t y, uint16_t color)
+{
+    if ((x < 0) || (x >= width()) || (y < 0) || (y >= height()))
+    {
+        return;
+    }
+
+    // check rotation, move pixel around if necessary
+    switch (getRotation())
```

Comment:
You can replace this whole switch using OOP with just one line:
_rotate(x, y, with(), height());